tips/TIP-0040/tip-0040.md

```diff
@@ -0,0 +1,848 @@
+  Delegation output on the output side.
+- Let `Past Bounded Slot Index` be given by `Commitment Index + Max Committable Age` where `Commitment Index` is the
+  slot index of the commitment input.
+- Let `Past Bounded Epoch Index` be the epoch index corresponding to the slot index `Past Bounded Slot Index`.
+- Let `Future Bounded Slot Index` be given by `Commitment Index + Min Committable Age` where `Commitment Index` is the
+  slot index of the commitment input.
+- Let `Future Bounded Epoch Index` be the epoch index corresponding to the slot index `Future Bounded Slot Index`.
+- If the past bounded indices are required for transaction validation, a _Commitment Input_ must be present. This is the
+  case for any transaction containing a _Delegation Output_ on the input or output side except when it is destroyed in
+  _Delayed Claiming State_.
+
+#### Consumed Outputs
+
+Whenever a Delegation output is consumed in a transaction, it means that the output is transitioned into its next state.
+From its initial state, the _Delegating State_, a Delegation Output can either be destroyed or transitioned to a
+_Delayed Claiming_ state, from which it can then be destroyed. The current state is defined as the consumed Delegation
+output, while the next state is defined as the Delegation output with the same explicit `Delegation ID` on the output
+side. There are thus two types of transitions: delayed claiming transition and destruction transition.
+
+- Delayed Claiming Transition
+  - Is identified by a zeroed-out `Delegation ID` on the input side and a non-zeroed-out `Delegation ID` on the output
+    side.
+  - The fields `Delegated Amount`, `Start Epoch` and `Validator Address` must not be changed.
+  - `End Epoch` on the output must be set as follows.
+    - Let `Registration Slot(n)` be defined as in [committee selection](#committee-selection), where `n` is equal to
+      `Future Bounded Epoch Index`.
+    - If `Future Bounded Slot Index <= Registration Slot(n+1)` then `End Epoch` must be set to
+      `Future Bounded Epoch Index`, otherwise to `Future Bounded Epoch Index + 1`.
+- Destruction Transition
+  - Is identified by the absence of a Delegation Output on the output side with an explicit `Delegation ID` that
+    corresponds to the `Delegation ID` of the one on the input side, which may be an implicit or explicit Delegation ID.
+    The next state is the empty state.
+  - During this transition, Mana rewards may be added to the total sum of Mana on the output side of the transaction,
+    according to [delegator rewards](#delegator-rewards).
+
+#### Created Outputs
+
+- When `Delegation ID` is zeroed out certain fields of the Delegation Output must be set as follows.
+  - Let `Registration Slot(n)` be defined as in [committee selection](#committee-selection), where `n` is equal to
+    `Past Bounded Epoch Index`.
+  - If `Past Bounded Slot Index <= Registration Slot(n+1)` then `Start Epoch` must be set to
```

Comment:
In the current implementation, this is implemented: `Past Bounded Slot Index <= Registration Slot(n)`, could you double-check?

https://github.com/iotaledger/iota.go/blob/c9fe55bd3b7c215b97002e59c332ea75b6dba981/vm/nova/vm.go#L979

Reply:
In the TIP Registration Slot(n+1) returns the slot within epoch n until which validators need to register for epoch n +1.
In the code, [registrationSlot(n)](https://github.com/iotaledger/iota.go/blob/c9fe55bd3b7c215b97002e59c332ea75b6dba981/vm/nova/vm.go#L1039-L1041) returns the slot within epoch n until which validators need to register for epoch n + 1.

So it's equivalent, just expressed slightly differently.

Reply:
Got it, thanks!!